translators/src/Print/Idris.hs

```diff
@@ -1,110 +1,155 @@
+  printExpr expr
+printExpr (If cond thn els) =
+  "if" <+> printExpr cond <+> "then" <+> printExpr thn <+> "else" <+> printExpr els
+printExpr (Where expr ds) =
+  printExpr expr <> line <>
+  indent 4 ("where" <> vcat (map printLocalDefn ds))
+printExpr (FunCall fun args) = pretty fun <+> (fillSep (map (group . printExpr) args))
+printExpr (VecE l) = encloseSep lbracket rbracket (comma <> space) (map printExpr l)
+printExpr (ListE l) = encloseSep lbracket rbracket (comma <> space) (map printExpr l)
+printExpr (Suc t) = parens $ "S" <+> printExpr t
+
+printOp :: Op -> Doc ann
+printOp Plus = "+"
+
+printLocalDefn :: LocalDefn -> Doc ann
+printLocalDefn (LocDefFun var ty args expr) = 
+  typeSig <> tvar <+> assign <+> align (printExpr expr)
+    where
+        typeSig = case ty of
+            Just t -> typeAnn (pretty var) (printType t) <> line
+            Nothing -> mempty
+        tvar = case args of
+            [] -> pretty var
+            (_:_) -> pretty var <+> (hsep $ map (pretty . arg) args)
+
+printDef :: Definition -> Doc ann
+printDef (DefTVar var Nothing expr) = 
+  pretty var <+> assign <+> align (printExpr expr) <> softline'
+printDef (DefTVar var (Just t) expr) = 
+  typeAnn (pretty var) (printType t) <> hardline <>
+  pretty var <+> assign <+> align (printExpr expr) <> hardline
+
+printDef (DefFun var ty args expr) = printLocalDefn (LocDefFun var ty args expr)
 printDef (DefPatt var params ty _ cons) =
-    var ++ typedel ++ printType (foldr Arr ty (map snd params)) ++
-    unwords (map (\(a,e) -> "\n" ++ var ++ " " ++ (unwords $ map arg a) ++ assign ++ printExpr e ) cons)
-printDef (DefDataType name cons ty) = "data " ++ name ++ typedel ++ printType ty ++ " where" ++
-    (line $ unwords (map (\(n, t) -> "\n " ++ n ++ typedel ++ printType t) cons))
-printDef (DefPDataType name params cons ty) = "data " ++ name ++ typedel ++
-    foldr (\(x, y) z -> parens (x ++ typedel ++ printType y) ++ arr ++ z) (printType ty) params ++ " where" ++ 
-    unwords (map (\(n, t) -> "\n " ++ n ++ typedel ++ line (intercalate arr (map fst params) ++ arr ++ printType t)) cons)
-
--- Record Defn
+    typeAnn (pretty var) (printType (foldr Arr ty (map snd params))) <> line <>
+    vsep (map (\(a, e) -> (pretty var) <+> hsep (map (pretty . arg) a) <+> assign <+> printExpr e) cons)
+printDef (DefDataType name cons ty) =
+  data_ <+> typeAnn (pretty name) (printType ty) <+> "where" <> hardline <>
+  indent 1 (vsep (map (\(n, t) -> typeAnn (pretty n) (printType t)) cons)) <>
+   hardline
+printDef (DefPDataType name params cons ty) =
+  data_ <+> 
+    typeAnn (pretty name) (concatWith (\x y -> x <+> arr <+> y) (map (\(x, y) -> teleCell (pretty x) (printType y)) params)) <+> arr <+> (printType ty) <+> 
+    "where" <> hardline <>
+    indent 1 (vsep (map (\(n,t) -> typeAnn (pretty n) 
+                                   (encloseSep emptyDoc (space <> arr) (space <> arr <> space) (map (pretty.fst) params)) <+> printType t) cons)) <>
+    hardline
```

Comment:
Wow that is a lot! Could we break this up into some `where` bindings so it looks something more like
```suggestion
printDef (DefPDataType name params cons ty) =
  data_ <+> typeAnn (pretty name) prettyParams  <+> arr <+> (printType ty) <+> "where <> hardline <>
  indent 1 (vsep (map prettyCon) cons) <> hardline
  where
    prettyParams = 
      concatWith (\x y -> x <+> arr <+> y) $
       map (\(x, y) -> teleCell (pretty x) (printType y)) params)
    prettyCon (n,t) = 
      typeAnn (pretty n) (encloseSep emptyDoc (space <> arr) (space <> arr <> space) (map (pretty . fst) params)) <+> printType t
```

Also I have no idea what is going on with the `encloseSep` here: why are we sticking arrows between all the datatype parameters??

Reply:
yes to some `where`. The arrows were these in the original pretty-printer and tests; I was just assuming that this was necessary Idris syntax! If it's not, I'd be quite happy to change this.

Reply:
According to the docs we dont:
https://idris2.readthedocs.io/en/latest/tutorial/typesfuns.html#vectors
Let's leave a NOTE or FIXME until we can get an end-to-end test running.